File: Legacy/Leetcode/79WordSearch.py
# Given a 2D board and a word, find if the word exists in the grid.
#
# The word can be constructed from letters of sequentially adjacent cell, where "adjacent" cells are those horizontally or vertically neighboring. The same letter cell may not be used more than once.
#
# For example,
# Given board =
#
# [
#   ['A','B','C','E'],
#   ['S','F','C','S'],
#   ['A','D','E','E']
# ]
# word = "ABCCED", -> returns true,
# word = "SEE", -> returns true,
# word = "ABCB", -> returns false.

def find_word(board, i, j, word):
    if len(word) == 0: # all the characters are checked
        return True
    if i<0 or i>=len(board) or j<0 or j>=len(board[0]) or word[0]!=board[i][j]:
        return False
    tmp = board[i][j]  # first character is found, check the remaining part
    board[i][j] = '#'  # avoid visit agian
    # check whether can find "word" along one direction
    res = find_word(board, i+1, j, word[1:]) or find_word(board, i-1, j, word[1:]) \
    or find_word(board, i, j+1, word[1:]) or find_word(board, i, j-1, word[1:])
    board[i][j] = tmp
    return res


def exist(board, word):
    if not board :
        return False
    for i in range(len(board)):
        for j in range(len(board[0])):
            if find_word(board, i, j, word):
                return True
    return False

board = [['A','B','C','E'],['S','F','C','S'],['A','D','E','E']]
word1 = "ABCCED"
word2 = "SEE"
word3 = "ABCB"
board2 = ["ABCE","SFES","ADEE"]
board2 = [[ele for ele in row] for row in board2]
word4 = 'ABCESEEEFS'

print exist(board, word1)
print exist(board, word2)
print exist(board, word3)
print exist(board2, word4)

File: C++/Backtracking/79. Word Search.cpp
class Solution {
    
    int m;
    int n;
    vector<pair<int, int>> adjs = {{-1,0},{1,0},{0,-1},{0,1}};
    function<bool(int, int)> outbound = [&](int i, int j)->bool{return i<0 || i>=m || j<0 || j>=n;};
    
public:
    
    bool backtrack(vector<vector<char>>& board, string& word, int i, int j, int index) {
        if (index == word.size()) {
            return true;
        }
        else if (outbound(i, j) || board[i][j] != word[index]) {
            return false;
        }
        else {
            board[i][j] = '#';
            bool res = false;
            for (auto [di, dj] : adjs) {
                int ni = i + di, nj = j + dj;
                res = res || backtrack(board, word, ni, nj, index + 1);
            }
            board[i][j] = word[index];
            return res;
        }
    }
    
    bool exist(vector<vector<char>>& board, string word) {
        m = board.size();
        n = board[0].size();
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (backtrack(board, word, i, j, 0)) return true;
            }
        }
        return false;
    }
};